File: packages/js/core/src/manifest/validators.ts
import { Uri } from "../types";

export function file(path: unknown): boolean {
  if (typeof path !== "string") {
    return false;
  }

  // eslint-disable-next-line no-useless-escape
  const validPathMatch = path.match(/^((\.\/|..\/)[^\/ ]*)+\/?$/gm);

  let result = false;
  if (validPathMatch && validPathMatch[0]) {
    result = validPathMatch[0].length === path.length;
  }

  return result;
}

export function packageName(name: unknown): boolean {
  return typeof name === "string";
}

export function packageTag(tag: unknown): boolean {
  return typeof tag === "string";
}

export function dockerImageName(name: unknown): boolean {
  return typeof name === "string";
}

export function dockerfileName(value: unknown): boolean {
  return (
    typeof value === "string" && file(value) && value.indexOf("Dockerfile") > -1
  );
}

export function dockerImageId(value: unknown): boolean {
  return typeof value === "string" && value.indexOf("sha256:") > -1;
}

export function wasmLanguage(language: unknown): boolean {
  return (
    typeof language === "string" &&
    (language === "interface" || language.indexOf("wasm/") > -1)
  );
}

export function pluginLanguage(language: unknown): boolean {
  return typeof language === "string" && language.indexOf("plugin/") > -1;
}

export function appLanguage(language: unknown): boolean {
  return typeof language === "string" && language.indexOf("app/") > -1;
}

export function imageFile(filePath: unknown): boolean {
  return (
    typeof filePath === "string" &&
    file(filePath) &&
    !!filePath.match(/(\.svg|\.png)$/)?.length
  );
}

export function websiteUrl(url: unknown): boolean {
  function validUrl(str: string) {
    const pattern = new RegExp(
      "^(https?:\\/\\/)?" + // protocol
        "((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|" + // domain name
        "((\\d{1,3}\\.){3}\\d{1,3}))" + // OR ip (v4) address
        "(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*" + // port and path
        "(\\?[;&a-z\\d%_.~+=-]*)?" + // query string
        "(\\#[-a-z\\d_]*)?$", // fragment locator
      "i"
    );
    return !!pattern.test(str);
  }
  return typeof url === "string" && validUrl(url);
}

export function graphqlFile(filePath: unknown): boolean {
  return (
    typeof filePath === "string" &&
    file(filePath) &&
    !!filePath.match(/(\.graphql)$/)?.length
  );
}

export function jsonFile(filePath: unknown): boolean {
  return (
    typeof filePath === "string" &&
    file(filePath) &&
    !!filePath.match(/(\.json)$/)?.length
  );
}

export function yamlFile(filePath: unknown): boolean {
  return (
    typeof filePath === "string" &&
    file(filePath) &&
    !!filePath.match(/(\.yaml)$/)?.length
  );
}

export function manifestFile(filePath: unknown): boolean {
  return jsonFile(filePath) || yamlFile(filePath);
}

export function regexString(regex: unknown): boolean {
  if (typeof regex !== "string") {
    return false;
  }

  let isValid = true;
  try {
    new RegExp(regex);
  } catch (e) {
    isValid = false;
  }

  return isValid;
}

export function polywrapUri(uri: unknown): boolean {
  return typeof uri === "string" && Uri.isValidUri(uri);
}

export function schemaFile(filePath: unknown): boolean {
  return typeof filePath === "string" && file(filePath);
}

export function directory(path: unknown): boolean {
  if (typeof path === "boolean") {
    return true;
  }
  if (typeof path !== "string") {
    return false;
  }

  const validDirRegex = /^\/?[\w\-/]+$/;

  return !!validDirRegex.test(path);
}

File: packages/py/manifests/polywrap/src/validators.py
import re
from typing import Any, Final, Literal, Pattern, TypeGuard, Union

DIRECTORY_PATTERN: Final[Pattern] = re.compile(r'^/?[\w\-/]+$')
IMAGE_PATTERN: Final[Pattern] = re.compile(r'(\.svg|\.png)$')
PATH_PATTERN: Final[Pattern] = re.compile(r'^((\./|../)[^/ ]*)+/?$', re.M)
URL_PATTERN: Final[Pattern] = re.compile(
    r'^(https?://)?((([a-z\d]([a-z\d-]*[a-z\d])*)\.)+[a-z]{2,}|((\d{1,3}\.){3}\d{1,3}))(:\d+)?(/[-a-z\d%_.~+]*)*(\?[;&a-z\d%_.~+=-]*)?(#[-a-z\d_]*)?$',
    re.I,
)


def file(path: Any) -> TypeGuard[str]:
    if not isinstance(path, str):
        return False

    result = re.search(PATH_PATTERN, path, re.M)
    return result is not None and result.group() == path


def package_name(name: Any) -> TypeGuard[str]:
    return isinstance(name, str)


def package_tag(tag: Any) -> TypeGuard[str]:
    return isinstance(tag, str)


def docker_image_name(name: Any) -> TypeGuard[str]:
    return isinstance(name, str)


def dockerfile_name(value: Any) -> TypeGuard[str]:
    return isinstance(value, str) and file(value) and 'Dockerfile' in value


def docker_image_id(value: Any) -> TypeGuard[str]:
    return isinstance(value, str) and 'sha256:' in value


def wasm_language(language: Any) -> TypeGuard[Union[Literal['interface'], str]]:
    return isinstance(language, str) and (language == 'interface' or 'wasm/' in language)


def plugin_language(language: Any) -> TypeGuard[str]:
    return isinstance(language, str) and 'plugin/' in language


def app_language(language: Any) -> TypeGuard[str]:
    return isinstance(language, str) and 'app/' in language


def image_file(file_path: Any) -> TypeGuard[str]:
    return file(file_path) and re.search(IMAGE_PATTERN, file_path) is not None


def website_url(url: Any) -> TypeGuard[str]:
    return isinstance(url, str) and re.fullmatch(URL_PATTERN, url, re.I) is not None


def graphql_file(file_path: Any) -> TypeGuard[str]:
    return file(file_path) and file_path.endswith('.graphql')


def json_file(file_path: Any) -> TypeGuard[str]:
    return file(file_path) and file_path.endswith('.json')


def yaml_file(file_path: Any) -> TypeGuard[str]:
    return file(file_path) and file_path.endswith('.yaml')


def manifest_file(filePath: Any) -> TypeGuard[str]:
    return json_file(filePath) or yaml_file(filePath)


def regex_string(regex: Any) -> TypeGuard[str]:
    try:
        re.compile(regex)
        return True
    except:
        return False


def web3_api_uri(uri: Any) -> TypeGuard[Uri]:
    return isinstance(uri, str) and Uri.is_valid_uri(uri)


def schema_file(file_path: Any) -> TypeGuard[str]:
    return file(file_path)


def directory(path: Any) -> TypeGuard[Union[bool, str]]:
    return isinstance(path, bool) or (isinstance(path, str) and re.fullmatch(DIRECTORY_PATTERN, path) is not None)


def buildx_output(output: Any) -> TypeGuard[Union[bool, str]]:
    return isinstance(output, bool) or (isinstance(output, str) and output in {'docker', 'registry'})
